### PR TITLE
SAK-40009 - Instructor name and id appears in student list when grading a 'Group …

### PR DESCRIPTION
**Pull request by SedueRey**
_…Submission - One submission per group' assignment_

This regression has shown up has shown some side effects problems between Gradebook 2 and Assignment because tries to put a grade on users which are not allowed to submit an assignment.

### CODE REVIEW

**Comment by juanjmerono** [comment]
I think this could be improved using streams and removing the if inside the loop, something like:

```
for (Member member : group.getMembers().stream().filter(m -> m.getRole().isAllowed(SECURE_ADD_ASSIGNMENT_SUBMISSION)).collect(Collectors.toList())) {

....

}

```
Or even:

```
group.getMembers().stream()
   .filter(m -> m.getRole().isAllowed(SECURE_ADD_ASSIGNMENT_SUBMISSION))
   .forEach(member -> {
       ......
    });

```

**Comment by SedueRey** [comment]
Sure. I'm going to update it right now

**Comment by SedueRey** [comment]
I've made those changes. Thanks, @juanjmerono

**Comment by mpellicer** [comment]
What's the Jira number?

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java [hunk 1]
```diff
@@ -996,12 +996,14 @@ public AssignmentSubmission addSubmission(String assignmentId, String submitter)
...
                     Group group = site.getGroup(submitter);
                     if (group != null && assignment.getGroups().contains(group.getReference())) {
-                        for (Member member : group.getMembers()) {
-                            AssignmentSubmissionSubmitter ass = new AssignmentSubmissionSubmitter();
-                            ass.setSubmitter(member.getUserId());
-                            submissionSubmitters.add(ass);
-                        }
-                        submission.setGroupId(submitter);
+                    	group.getMembers().stream()
+                    	.filter(m -> m.getRole().isAllowed(SECURE_ADD_ASSIGNMENT_SUBMISSION))
```
Comment (mpellicer) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java]: I think this checks the permissions at a site level, and not at a group level. An instructor could have this permission disabled at site level but enabled at a group level.

**Comment by SedueRey** [comment]
I've update the PR name (the jira is SAK-40009)

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java [hunk 2]
```diff
@@ -1536,12 +1538,15 @@ public String getSubmissionStatus(String submissionId) {
...
     public List<User> getSortedGroupUsers(Group g) {
         List<User> users = new ArrayList<>();
-        for (Member member : g.getMembers())
-            try {
-                users.add(userDirectoryService.getUser(member.getUserId()));
-            } catch (Exception e) {
-                log.warn("Creating a list of users, user = {}, {}", member.getUserId(), e.getMessage());
-            }
+		g.getMembers().stream()
+		.filter(m -> m.getRole().isAllowed(SECURE_ADD_ASSIGNMENT_SUBMISSION))
```
Comment (mpellicer) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java]: Same as before, probably is better to check at group level.

File: assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java [hunk 3]
```diff
@@ -690,6 +691,9 @@ private AssignmentSubmission createNewGroupSubmission(String context, String gro
         submitters.forEach(s -> {
                 Member member = mock(Member.class);
                 when(member.getUserId()).thenReturn(s);
+                Role r = mock(Role.class);
+                when(member.getRole()).thenReturn(r);
+                when(r.isAllowed(AssignmentServiceConstants.SECURE_ADD_ASSIGNMENT_SUBMISSION)).thenReturn(true);
```
Comment (mpellicer) [inline assignment/impl/src/test/org/sakaiproject/assignment/impl/AssignmentServiceTest.java]: Same

**Comment by SedueRey** [comment]
I've updated the PR , @mpellicer to check what you said but I think @juanjmerono disagree a little ;)

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java [hunk 4]
```diff
@@ -996,12 +996,14 @@ public AssignmentSubmission addSubmission(String assignmentId, String submitter)
...
                     Group group = site.getGroup(submitter);
                     if (group != null && assignment.getGroups().contains(group.getReference())) {
-                        for (Member member : group.getMembers()) {
-                            AssignmentSubmissionSubmitter ass = new AssignmentSubmissionSubmitter();
-                            ass.setSubmitter(member.getUserId());
-                            submissionSubmitters.add(ass);
-                        }
-                        submission.setGroupId(submitter);
+                    	group.getMembers().stream()
+                    	.filter(m -> m.getRole().isAllowed(SECURE_ADD_ASSIGNMENT_SUBMISSION) || group.isAllowed(m.getUserId(), SECURE_ADD_ASSIGNMENT_SUBMISSION) )
```
Comment (juanjmerono) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java]: I think both checks seems to be the same (group isAllowed is calling role.isAllowed): 

https://github.com/sakaiproject/sakai/blob/master/kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/BaseAuthzGroup.java#L762-L784

Do you think there are use cases that return different values @mpellicer ?

Comment (mpellicer) [reply assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java]: Some permissions of Assignments are checked at a site level and at a group level, makes more sense to me to check at group level here.

Just saying, every time I worked with site and group permissions I faced this stuff, it's a bit messy.

Have you tested how it works after the changes @SedueRey ? Is it working for both cases? Is it working like the previous PR without the OR comparison?

**Comment by juanjmerono** [comment]
@SedueRey I've seen a little contradiction here: 
The addSubmission method is called in the instructor's first access to the submissions, so the instructor must have that permission according to:
https://github.com/sakaiproject/sakai/blob/master/assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java#L980-L988

But you are supposing that only students has this permission. I think you should check other permission in your filter to discard instructors in the list something like SECURE_ADD_ASSIGNMENT could be more appropiate.

**Comment by SedueRey** [comment]
Sure. I've though that this permission was the appropriate so great catch, I'll update this next monday so if you could add wip tag it would be great

**Comment by SedueRey** [comment]
I've updated the permission check but I think @ern should take a look on this modifications so I'll remove WIP after Core Team call to discuss it

**Comment by ern** [comment]
This will typically filter out Instructors, and TA's who are in the group.
Looks good as it its checking permissions.

**Comment by SedueRey** [comment]
Thanks, @ern , I really appreciate

**Pull request merged**